zerver/lib/user_groups.py

```diff
@@ -325,15 +325,29 @@ def get_recursive_subgroups(user_group: UserGroup) -> QuerySet[UserGroup]:
     cte = With.recursive(
         lambda cte: UserGroup.objects.filter(id=user_group.id)
         .values(group_id=F("id"))
-        .union(cte.join(UserGroup, direct_supergroups=cte.col.group_id).values(group_id=F("id")))
+        .union(
+            cte.join(NamedUserGroup, direct_supergroups=cte.col.group_id).values(group_id=F("id"))
+        )
     )
     return cte.join(UserGroup, id=cte.col.group_id).with_cte(cte)
 
 
-def get_recursive_strict_subgroups(user_group: UserGroup) -> QuerySet[UserGroup]:
+def get_recursive_strict_subgroups(user_group: UserGroup) -> QuerySet[NamedUserGroup]:
     # Same as get_recursive_subgroups but does not include the
     # user_group passed.
-    return get_recursive_subgroups(user_group).exclude(id=user_group.id)
+    cte = With.recursive(
+        lambda cte: UserGroup.objects.filter(id=user_group.id)
+        .values(group_id=F("id"))
+        .union(
+            cte.join(NamedUserGroup, direct_supergroups=cte.col.group_id).values(group_id=F("id"))
+        )
+    )
+    # We can just query for NamedUserGroup here instead of UserGroup here as
+    # we only NamedUserGroup can be subgroups of a group and we do not want
+    # to include user_group passed to the function in the result.
+    # "exclude" condition is still required since the user_group passed to
+    # this function can be linked to a NamedUserGroup.
+    return cte.join(NamedUserGroup, id=cte.col.group_id).with_cte(cte).exclude(id=user_group.id)
```

Comment:
I don't think we can modify the recursive query to not include the group itself, so had to add a separate `exclude` clause.

Reply:
I think if you start with just `lambda cte: NamedUserGroup.objects.filter(id__in=user_group.direct_subgroups)`, you don't need to do the exclude.

Reply:
Yeah, sounds good, we can track that as a follow-up.